### ISSUE TITLE
Cannot roll initiative from the combat tracker

### ISSUE THREAD

**Issue by PaulWRoberts** [trioderegion/dnd5e-helpers#130]
In foundry 8.6.0 with the 2.0.5 DnD5e Helpers I get the following error when I try to roll initiative

```
Uncaught (in promise) Error: The key l5pyujNFm8YHwlUC does not exist in the EmbeddedCollection Collection
    at Map.get (/C:/Program%20Files/FoundryVTT/resources/app/common/utils/collection.mjs:106)
    at ServerDatabaseBackend._updateEmbeddedDocuments (/C:/Program%20Files/FoundryVTT/resources/app/dist/database/backend/server-backend.mjs:1)
    at ServerDatabaseBackend.update (/C:/Program%20Files/FoundryVTT/resources/app/common/abstract/backend.mjs:153)
    at async Socket.handleEvent (/C:/Program%20Files/FoundryVTT/resources/app/dist/server/sockets.mjs:1)
```

**Comment by trioderegion** [comment]
I am unable to reproduce this. Can you provide more specific reproduction steps?

**Comment by PaulWRoberts** [comment]
To replicate you need to press the roll all or roll npc buttons
![image](https://user-images.githubusercontent.com/5503568/121943079-0efea780-cd06-11eb-97d4-4df11f790b12.png)

**Comment by trioderegion** [comment]
I'm sorry, I still cannot reproduce. Are all other module disabled?

**Comment by PaulWRoberts** [comment]
I upgraded to 2.0.6 DnD5e Helpers, I am running 8.60 Foundry and on Windows 10. I made a new world and I only enabled DnD5e Helpers, no other mod. I created a new map, **which I set to gridless**. I added the example hero from the compendium "Akra" and an "Acolyte", I then added them both to the combat tracker by highlighting both, right clicking and pressing the toggle combat state button. Then before starting the encounter, I try pressing the Roll All and Roll NPCs buttons (pressing the dice works as expected) and I can reproduce 100% of the time.

**Comment by PaulWRoberts** [comment]
Ok, after some experimentation I found out what a crutial step for reproduction is.
If I add each of the 2 tokens to the tracker at the same time (select them both and toggle combat state) I reproduce.
If I add each of them one at a time (select one, add it and then select the other to add it) I cannot reproduce.

**Comment by PaulWRoberts** [comment]
After deleting the tokens from the scene and clearing the combat tracker I can no longer place the same tokens in the scene.

**Comment by trioderegion** [comment]
This is very strange and I'm starting to suspect is not due to helpers -- particularly the token placement issue.  Not trying to pawn you off at all, but the issue seems to be something more fundamental.

I havent tried gridless, so I'll attempt to reproduce that way.

**Comment by PaulWRoberts** [comment]
This is so bizarre, looking at the logs the error happens in the server code. I definitely can only reproduce the error with 5e helpers enabled. I did find a work around while trying to even figure out where to set a breakpoint in the js.

The work around is, if the initiative breaks, press the reset initiative button. I can roll initiative after that.

**Comment by thatlonelybugbear** [comment]
For me the same error almost is triggered by using a spell template that is concentration that bestows a CUB condition via a DAE on targets as well as the spellcaster (casts silence on a 20ft-radius that includes the caster and applies a deafened condition.

When the concentration ends or the time is up and the effects are deleted from the other targets, the error fires up and the effect cannot be deleted from the caster (either by going to the effects tab of the character sheet or the Assign Status Effects icon). 

The effect needs a Foundry reload to clear.

`Foundry VTT | Deleted ActiveEffect with id [UjMl1gCgZk6kDUnZ] in parent Actor [V2bDa2t6Yqj59D4E]
15:31:16.695 backend.mjs:306 Foundry VTT | Deleted ActiveEffect with id [VFxrTecwpTn413RT] in parent Actor [XCsuQ4q5F33R2fuO]
15:31:16.710 backend.mjs:306 Foundry VTT | Deleted ActiveEffect with id [Aj1LzOg4nBs6cYKc] in parent Actor [PwyTMcndlay7IuFm]
15:31:16.723 backend.mjs:306 Foundry VTT | Deleted ActiveEffect with id [Ive8jfXUf30l70gr] in parent Actor [Ie6WAcmUf6qHLWYp]
15:31:16.744 backend.mjs:306 Foundry VTT | Deleted ActiveEffect with id [ZnZdGkH5Zf7Zx7Gp] in parent Actor [Q8VBlroKgDXt9gxU]
15:31:16.784 backend.mjs:306 Foundry VTT | Deleted ActiveEffect with id [jPBdoPb1RvESGKws] in parent Actor [WhZdlNavtph5mbNt]
15:31:16.832 backend.mjs:306 Foundry VTT | Deleted ActiveEffect with id [4JnkpwY2g2sPaVPi] in parent Actor [gmgpjNPOHZKRgiRb]
15:31:16.843 GMAction.js:210 delete effects failed  Error: The key rnmRKGwTWz7LOIaL does not exist in the EmbeddedCollection Collection
    at Map.get (/C:/Program%20Files/FoundryVTT/resources/app/common/utils/collection.mjs:106)
    at ServerDatabaseBackend._deleteEmbeddedDocuments (/C:/Program%20Files/FoundryVTT/resources/app/dist/database/backend/server-backend.mjs:1)
    at ServerDatabaseBackend.delete (/C:/Program%20Files/FoundryVTT/resources/app/common/abstract/backend.mjs:214)
    at async Socket.handleEvent (/C:/Program%20Files/FoundryVTT/resources/app/dist/server/sockets.mjs:1)
deleteEffects @ GMAction.js:210
async function (async)
deleteEffects @ GMAction.js:207
processSingleAction @ GMAction.js:154
processAction @ GMAction.js:53
requestGMAction @ GMAction.js:14
handleRemoveConcentration @ dae.js:755
async function (async)
handleRemoveConcentration @ dae.js:753
update @ dae.js:826
_call @ VM41291:4
callAll @ foundry.js:153
callback @ foundry.js:9048
(anonymous) @ foundry.js:9019
_handleDeleteEmbeddedDocuments @ foundry.js:9019
_deleteEmbeddedDocuments @ foundry.js:8928
async function (async)
_deleteEmbeddedDocuments @ foundry.js:8917
delete @ backend.mjs:214
async function (async)
delete @ backend.mjs:212
deleteDocuments @ document.mjs:401
delete @ document.mjs:462
_toggleActiveEffect @ foundry.js:45164
toggleEffect @ foundry.js:45132
_onToggleEffect @ foundry.js:30781
dispatch @ jquery.min.js:2
v.handle @ jquery.min.js:2
15:31:16.855 GMAction.js:210 delete effects failed  Error: The key rnmRKGwTWz7LOIaL does not exist in the EmbeddedCollection Collection
    at Map.get (/C:/Program%20Files/FoundryVTT/resources/app/common/utils/collection.mjs:106)
    at ServerDatabaseBackend._deleteEmbeddedDocuments (/C:/Program%20Files/FoundryVTT/resources/app/dist/database/backend/server-backend.mjs:1)
    at ServerDatabaseBackend.delete (/C:/Program%20Files/FoundryVTT/resources/app/common/abstract/backend.mjs:214)
    at async Socket.handleEvent (/C:/Program%20Files/FoundryVTT/resources/app/dist/server/sockets.mjs:1)
deleteEffects @ GMAction.js:210
async function (async)
deleteEffects @ GMAction.js:207
processSingleAction @ GMAction.js:154
processAction @ GMAction.js:53
requestGMAction @ GMAction.js:14
handleRemoveConcentration @ dae.js:755
async function (async)
handleRemoveConcentration @ dae.js:753
update @ dae.js:826
_call @ VM41291:4
callAll @ foundry.js:153
callback @ foundry.js:9048
(anonymous) @ foundry.js:9019
_handleDeleteEmbeddedDocuments @ foundry.js:9019
_deleteEmbeddedDocuments @ foundry.js:8928
async function (async)
_deleteEmbeddedDocuments @ foundry.js:8917
delete @ backend.mjs:214
async function (async)
delete @ backend.mjs:212
deleteDocuments @ document.mjs:401
delete @ document.mjs:462
_toggleActiveEffect @ foundry.js:45164
toggleEffect @ foundry.js:45132
_onToggleEffect @ foundry.js:30781
dispatch @ jquery.min.js:2
v.handle @ jquery.min.js:2
15:31:22.566 foundry.js:1246 Uncaught (in promise) Error: The key rnmRKGwTWz7LOIaL does not exist in the EmbeddedCollection Collection
    at Map.get (/C:/Program%20Files/FoundryVTT/resources/app/common/utils/collection.mjs:106)
    at ServerDatabaseBackend._deleteEmbeddedDocuments (/C:/Program%20Files/FoundryVTT/resources/app/dist/database/backend/server-backend.mjs:1)
    at ServerDatabaseBackend.delete (/C:/Program%20Files/FoundryVTT/resources/app/common/abstract/backend.mjs:214)
    at async Socket.handleEvent (/C:/Program%20Files/FoundryVTT/resources/app/dist/server/sockets.mjs:1)`

**Comment by thatlonelybugbear** [comment]
Sorry have to clarify. 

The initiative bug happens to me from time to time, but cannot reproduce effectively. That's why I found this thread!

But I can reproduce the same error message as the OP by the steps provided above.

**Comment by thatlonelybugbear** [comment]
OK this is beginning to drive me insane ^^

Now that I reloaded a couple of times Foundry and re-enabled DnD5e Helpers I cannot reproduce anymore...

Maybe disabling dnd5e helpers and trying it out again after was what changed the behaviour...

**Comment by thatlonelybugbear** [comment]
Update: New world and only dnd5e helpers enabled.

https://1drv.ms/v/s!As4ozpu6lMUZj4U8PrMmn3zqlZfpfw?e=vui5yT

**Comment by thatlonelybugbear** [comment]
Template issue?

https://1drv.ms/u/s!As4ozpu6lMUZj4U9oG646bb9hHUJRQ?e=8WPiXY

**Comment by vanderleitelesjr** [comment]
Same issue here
Rolling initiative for the first combat in the game goes just fine, but after ending that combat and creating a new one the error begins to pop up.
For testing purpose, it does happen with significant greater frequency if there is another player logged in the game, testing using only the GM player does seem to suppress this a little bit.

**Comment by trioderegion** [comment]
> Same issue here
> Rolling initiative for the first combat in the game goes just fine, but after ending that combat and creating a new one the error begins to pop up.
> For testing purpose, it does happen with significant greater frequency if there is another player logged in the game, testing using only the GM player does seem to suppress this a little bit.

Does this happen more often if you have creatures targeted when doing these actions? or is this without any targeting going on?

I'll be looking into this issue tomorrow trying to resolve it.

**Comment by vanderleitelesjr** [comment]
no targets, just tested with one of my players
tried to simply create a combat, roll init, end it and create a new combat, the issue then pops up when rolling init for the second combat onwards
no tokens selected as targets
disabled helpers and the issue did not return
tried to re-enable it and test each setting from the module on and off, but the issue persisted

**Comment by vanderleitelesjr** [comment]
no targets on my end, I have to say, did not notice if my player had anything targeted

**Comment by PaulWRoberts** [comment]
No targets and only one GM account and no player accounts logged in. This does seem to be a bit racy as well, and as I said before has a greater chance of occurring when multiple combatants are entered at once.

I looked through your code and I did see that there were a lot of awaits into a PIXI library in createCombatant. Perhaps the delay this causes introduces some delay that causes the client and the server to be out of sync. The errors do indicate that the server disagrees that these combatants are in the combat.

**Comment by trioderegion** [comment]
I have greatly reduced the chance for these deletion race conditions to occur, though I can't *quite* pin down the root cause. These fixes will be coming out with 2.0.7 and I would greatly appreciate trying to break it again with as specific reproduction steps as possible.

I am closing this issue for now. If the error still exists, please file a new issue.

@thatlonelybugbear The template targeting error video you posted is extremely difficult for me to reproduce as you have a number of other modules enabled that I do not have any experience with.  If you can reproduce that same error when 2.0.7 releases, I would appreciate a new issue filed with **as few modules enabled as possible** that still showcases the problem.
